ceph/crashes.go

```diff
@@ -0,0 +1,152 @@
+//   Copyright 2022 DigitalOcean
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package ceph
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"regexp"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	crashLsLineRegex = regexp.MustCompile(`.*_[0-9a-f-]{36}\s+(\S+)\s*(\*)?`)
+
+	statusNames = map[bool]string{true: "new", false: "archived"}
+)
+
+// CrashesCollector collects information on how many crash reports are currently open.
+// These reports are counted by daemon/client name, and by status (new or archived).
+// This is NOT the same as new_crash_reports, that only counts new reports in the past
+// two weeks as reported by 'ceph health'.
+type CrashesCollector struct {
+	conn    Conn
+	logger  *logrus.Logger
+	version *Version
+
+	// We keep track of which daemons we've seen so that their error count
+	// can be reset to zero if the errors get purged.
+	knownEntities map[string]bool
+
+	CrashReports prometheus.GaugeVec
+}
+
+// NewCrashesCollector creates a new CrashesCollector instance
+func NewCrashesCollector(exporter *Exporter) *CrashesCollector {
+	labels := make(prometheus.Labels)
+	labels["cluster"] = exporter.Cluster
+
+	collector := &CrashesCollector{
+		conn:    exporter.Conn,
+		logger:  exporter.Logger,
+		version: exporter.Version,
+
+		knownEntities: map[string]bool{},
+
+		CrashReports: *prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace:   cephNamespace,
+				Name:        "crash_reports",
+				Help:        "Count of crashes reports per daemon, according to `ceph crash ls`",
+				ConstLabels: labels,
+			},
+			[]string{"daemon", "status"},
+		),
+	}
+
+	return collector
+}
+
+type crashEntry struct {
+	entity string
+	isNew  bool
+}
+
+// getCrashLs runs the 'crash ls' command and parses its results
+func (c *CrashesCollector) getCrashLs() ([]crashEntry, error) {
+	crashes := make([]crashEntry, 0)
+
+	// We parse the plain format because it is quite compact.
+	// The JSON output of this command is very verbose and might be too slow
+	// to process in an outage storm.
```

Comment:
So, the downside to this approach is that the json format should hopefully be a bit more stable. I can understand your hesitation to go down the json path, though.

Reply:
Maybe I'm too cautious yeah. Looks like 2000 errors can create around 7.5MB of JSON, and I suspect we might see higher sizes. Is that a lot or is that reasonable?

Reply:
That doesn't sound too bad to me, and hopefully we (and others) aren't letting the list get that long.